### PR TITLE
executor: make sure hashjoin's goroutine exit before `Close` return

### LINKED ISSUE #8337: hash-join executor close returned but forked goroutine still running
## Bug Report

1. What did you do?
If possible, provide a recipe for reproducing the error.

run CI

2. What did you expect to see?

CI green

3. What did you see instead?

found data race:

https://internal.pingcap.net/jenkins/blue/organizations/jenkins/tidb_ghpr_unit_test/detail/tidb_ghpr_unit_test/4327/

```
==================
WARNING: DATA RACE
Write at 0x00c000b67090 by goroutine 388:
  github.com/pingcap/tidb/executor.(*tableResultHandler).Close()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/table_reader.go:199 +0x15a
  github.com/pingcap/tidb/executor.(*TableReaderExecutor).Close()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/table_reader.go:117 +0xc2
  github.com/pingcap/tidb/executor.(*baseExecutor).Close()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/executor.go:99 +0x97
  github.com/pingcap/tidb/executor.(*HashJoinExec).Close()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join.go:133 +0x49b
  github.com/pingcap/tidb/executor.(*baseExecutor).Close()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/executor.go:99 +0x97
  github.com/pingcap/tidb/executor.(*SelectionExec).Close()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/executor.go:835 +0xa8
  github.com/pingcap/tidb/executor.(*baseExecutor).Close()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/executor.go:99 +0x97
  github.com/pingcap/tidb/executor.(*ProjectionExec).Close()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/projection.go:246 +0x136
  github.com/pingcap/tidb/executor.(*ExplainExec).generateExplainInfo()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/explain.go:84 +0x2fe
  github.com/pingcap/tidb/executor.(*ExplainExec).Next()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/explain.go:51 +0x9a3
  github.com/pingcap/tidb/executor.(*recordSet).Next()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/adapter.go:96 +0x97
  github.com/pingcap/tidb/session.GetRows4Test()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/tidb.go:216 +0x39b
  github.com/pingcap/tidb/util/testkit.(*TestKit).ResultSetToResult()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:186 +0x13e
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustQuery()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:180 +0x4f2
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustQuery()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:177 +0x259
  github.com/pingcap/tidb/executor_test.(*testSuite).TestHashJoin()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join_test.go:953 +0x990
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustQuery()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:177 +0x259
  github.com/pingcap/tidb/executor_test.(*testSuite).TestHashJoin()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join_test.go:952 +0x8dc
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustQuery()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:177 +0x259
  github.com/pingcap/tidb/executor_test.(*testSuite).TestHashJoin()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join_test.go:951 +0x828
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustExec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:166 +0x94
  github.com/pingcap/tidb/executor_test.(*testSuite).TestHashJoin()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join_test.go:950 +0x7f1
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustExec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:166 +0x94
  github.com/pingcap/tidb/executor_test.(*testSuite).TestHashJoin()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join_test.go:949 +0x7ba
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustQuery()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:177 +0x259
  github.com/pingcap/tidb/executor_test.(*testSuite).TestHashJoin()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join_test.go:935 +0x3aa
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustExec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:166 +0x94
  github.com/pingcap/tidb/executor_test.(*testSuite).TestHashJoin()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join_test.go:934 +0x373
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustQuery()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:177 +0x259
  github.com/pingcap/tidb/executor_test.(*testSuite).TestHashJoin()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join_test.go:933 +0x2bf
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustQuery()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:177 +0x259
  github.com/pingcap/tidb/executor_test.(*testSuite).TestHashJoin()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join_test.go:932 +0x20b
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustExec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:166 +0x94
  github.com/pingcap/tidb/executor_test.(*testSuite).TestHashJoin()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join_test.go:931 +0x1d4
  github.com/pingcap/tidb/session.(*session).execute()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:815 +0x9d8
  github.com/pingcap/tidb/session.(*session).Execute()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:768 +0x84
  github.com/pingcap/tidb/util/testkit.(*TestKit).Exec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:137 +0x12d
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustExec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:166 +0x94
  github.com/pingcap/tidb/executor_test.(*testSuite).TestHashJoin()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join_test.go:930 +0x19d
  github.com/pingcap/tidb/session.(*session).execute()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:815 +0x9d8
  github.com/pingcap/tidb/session.(*session).Execute()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:768 +0x84
  github.com/pingcap/tidb/util/testkit.(*TestKit).Exec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:137 +0x12d
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustExec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:166 +0x94
  github.com/pingcap/tidb/executor_test.(*testSuite).TestHashJoin()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join_test.go:929 +0x166
  github.com/pingcap/tidb/util/testkit.(*TestKit).Exec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:137 +0x12d
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustExec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:166 +0x94
  github.com/pingcap/tidb/executor_test.(*testSuite).TestHashJoin()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join_test.go:928 +0x12f
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustExec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:166 +0x94
  github.com/pingcap/tidb/executor_test.(*testSuite).TestHashJoin()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join_test.go:927 +0xf8
  runtime.call32()
      /usr/local/go/src/runtime/asm_amd64.s:522 +0x3a
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:308 +0xc0
  github.com/pingcap/tidb/vendor/github.com/pingcap/check.(*suiteRunner).forkTest.func1()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/vendor/github.com/pingcap/check/check.go:798 +0xa04
  github.com/pingcap/tidb/vendor/github.com/pingcap/check.(*suiteRunner).forkCall.func1()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/vendor/github.com/pingcap/check/check.go:692 +0x89
Previous read at 0x00c000b67090 by goroutine 477:
  github.com/pingcap/tidb/executor.(*tableResultHandler).nextChunk()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/table_reader.go:176 +0x67
  github.com/pingcap/tidb/executor.(*TableReaderExecutor).Next()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/table_reader.go:107 +0xa6
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchInnerRows()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join.go:270 +0x6a6
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchInnerAndBuildHashTable.func1()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join.go:528 +0x65
  github.com/pingcap/tidb/util.WithRecovery()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/misc.go:73 +0x5a
Goroutine 388 (running) created at:
  github.com/pingcap/tidb/vendor/github.com/pingcap/check.(*suiteRunner).forkCall()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/vendor/github.com/pingcap/check/check.go:689 +0x4a5
  github.com/pingcap/tidb/vendor/github.com/pingcap/check.(*suiteRunner).forkTest()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/vendor/github.com/pingcap/check/check.go:780 +0x126
  github.com/pingcap/tidb/vendor/github.com/pingcap/check.(*suiteRunner).run()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/vendor/github.com/pingcap/check/check.go:631 +0x22c
  github.com/pingcap/tidb/vendor/github.com/pingcap/check.Run()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/vendor/github.com/pingcap/check/run.go:96 +0x5a
  github.com/pingcap/tidb/vendor/github.com/pingcap/check.RunAll()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/vendor/github.com/pingcap/check/run.go:88 +0x129
  github.com/pingcap/tidb/vendor/github.com/pingcap/check.TestingT()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/vendor/github.com/pingcap/check/run.go:76 +0x79b
  github.com/pingcap/tidb/executor_test.TestT()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/executor_test.go:74 +0xef
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162
Goroutine 477 (finished) created at:
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchInnerAndBuildHashTable()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join.go:528 +0x1a9
  github.com/pingcap/tidb/executor.(*HashJoinExec).Next.func2()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join.go:496 +0x53
  github.com/pingcap/tidb/util.WithRecovery()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/misc.go:73 +0x5a
==================
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
2.1.0.rc5
```

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

fixes #8337 

### What is changed and how it works?

also wait `fetchInnerRows` goroutine exit when found `finished` be setted.

### Check List 

Tests 

 - Old test

Code changes

 - Close logic

Side effects

 -  N/A

Related changes

 - Need to cherry-pick to the release branch

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8338)

### CODE REVIEW

File: executor/join.go [hunk 1]
```diff
@@ -587,6 +594,8 @@ func (e *HashJoinExec) buildHashTableForList(innerResultCh chan *chunk.Chunk) er
 	chkIdx := uint32(0)
 	for chk := range innerResultCh {
 		if e.finished.Load().(bool) {
+			// wait fetchInnerRows goroutine exit.
+			fetchInnerDone.Wait()
```
Comment (zz-jason) [inline executor/join.go]: how about moving this `Wait()` to the `Close()` function?

File: executor/join.go [hunk 2]
```diff
@@ -550,7 +556,11 @@ func (e *HashJoinExec) fetchInnerAndBuildHashTable(ctx context.Context) {
 	// innerResultCh transfer inner result chunk from inner fetch to build hash table.
 	innerResultCh := make(chan *chunk.Chunk, e.concurrency)
 	doneCh := make(chan struct{})
-	go util.WithRecovery(func() { e.fetchInnerRows(ctx, innerResultCh, doneCh) }, nil)
+	e.fetchInnerDone.Add(1)
+	go util.WithRecovery(
+		func() { e.fetchInnerRows(ctx, innerResultCh, doneCh) },
+		func(r interface{}) { e.finishFetchInnerRows(innerResultCh) },
```
Comment (zz-jason) [inline executor/join.go]: only `finishFetchInnerRows()` when the `fetchInnerRows()` goroutine is paniced?

Comment (lysu) [reply executor/join.go]: not~ it alway call finishFetchInnerRows but with `r ==nil`

Comment (zz-jason) [reply executor/join.go]: 😂 I was mislead by the `r interface{}` parameter in the function call.

Comment (zz-jason) [reply executor/join.go]: can we move the cleaning code into that function, and only do the recover work in this function?

**Pull request merged**